client/core/core.go

```diff
@@ -4563,16 +4565,64 @@ func handleMatchProofMsg(c *Core, dc *dexConnection, msg *msgjson.Message) error
 		return fmt.Errorf("match proof note unmarshal error: %w", err)
 	}
 
-	// Expire the epoch
-	dc.setEpoch(note.MarketID, note.Epoch+1)
+	// Expire previous epoch.
+	dc.tryUpdateEpoch(note.MarketID, note.Epoch+1)
 
 	book := dc.bookie(note.MarketID)
 	if book == nil {
 		return fmt.Errorf("no order book found with market id %q",
 			note.MarketID)
 	}
 
-	return book.ValidateMatchProof(note)
+	err = book.ValidateMatchProof(note)
+	if err != nil {
+		return fmt.Errorf("match proof validation failed: %w", err)
+	}
+
+	err = validateMatchCSum(note.Preimages, note.CSum, dc.trackedTrades())
+	if err != nil {
+		// Validation can fail either due to server trying to cheat, or
+		// client loosing (e.g. due to restarting, since we don't persistently store these at the moment)
+		// his csums for the trades involved.
+		//
+		// We'll just warn the user for now (maybe we should handle this in a stricter way in the future).
+		c.log.Warnf("validate match checksum for epoch %d, err: %v", note.Epoch, err)
+	}
+
+	return nil
+}
+
+// validateMatchCSum for each preimage in the match validates
+// commited checksums on corresponding orders (trades)
+// against csum for this match, making sure they are of the same value.
+//
+// This is necessary to ensure that for an order O
+// "server preimage request" and "O being matched" happen in the same epoch,
+// otherwise server could game match shuffle, if it sees O's preimage
+// before committing to csum for a match.
+func validateMatchCSum(preimages []order.Preimage, matchCSum dex.Bytes, trades []*trackedTrade) error {
+	for _, preimage := range preimages {
+		err := validateMatchCSumForPreimage(preimage, matchCSum, trades)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateMatchCSumForPreimage(preimage order.Preimage, matchCSum dex.Bytes, trades []*trackedTrade) error {
+	for _, trade := range trades {
+		if preimage == trade.preImg {
```

Comment:
Don't check every trade, only those in `OrderStatusEpoch` for which we've received a preimage request.

Reply:
fixed.

Reply:
I commented in https://github.com/decred/dcrdex/pull/1077#discussion_r627812050, but I actually think we have to look at all of our trades' preimages because OrderStatusEpoch orders can be from multiple epochs.  If trackedTrade knew which epoch index it was in, that would work too, alas it does not know.

Reply:
> If trackedTrade knew which epoch index it was in

As long as we have the `order.Prefix.ServerTime` and the `msgjson.Market.EpochLen`, we have the epoch index, and we can match against `msgjson.MatchProofNote.Epoch`.

Reply:
Yep, which can simplify the filter too.

```
epochIdx := encode.UnixMilliU(trade.Prefix().ServerTime) / trade.epochLen
if note.Epoch == epochIdx && note.MarketID == trade.mktID {
    // do stuff
}
```

Note the addition of the market ID comparison, which I had failed to notice before.